required/tools/multicol.dtx

```diff
@@ -4641,7 +4327,346 @@
+%    \begin{macrocode}
+        \unvcopy\partial@page
+        \process@cols \mult@firstbox { \unvcopy\count@ }
+        \unvcopy\mult@rightbox
+      }
+}
+%    \end{macrocode}
+%  \end{macro}
+%
+
+%  \begin{macro}{\@@_prepare_mark_reinserts:}
+%    
+%    When finishing a \mc{} environment, we have to return marks (in
+%    then boxed columns) to the current page so that they are
+%    available when that page is produced.
+%    This is what this helper does.
+%    \begin{macrocode}
+\cs_new_protected:Npn \@@_prepare_mark_reinserts: {
+  \@@_debug_marks:n
+     { \typeout{Marks:~ prepare~ for~ reinserting~ marks~ (multicol)} }
+%    \end{macrocode}
+%    We are only interested in the top, first, and last marks of each
+%    class regardless in which column they appeared, so we can copy
+%    all column boxes together and process them in one go. The result
+%    is returned in \cs{\detokenize{l_@@_first_marks_tl}} and
+%    \cs{\detokenize{l_@@_last_marks_tl}} in form of \cs{\detokenize{mark_insert:nn}} statements
+%    so we can later execute such token lists directly (and if there
+%    were no marks they will be empty).
+%    \begin{macrocode}
+  \__mark_get_marks_for_reinsertion:nNN
+     {
+       \process@cols \mult@firstbox { \unvcopy\count@ }
+       \unvcopy\mult@rightbox
+     }
+     \l_@@_first_marks_tl
+     \l_@@_last_marks_tl
+}
+%    \end{macrocode}
+%    And here are the token lists used above.
+%    \begin{macrocode}
+\tl_new:N \l_@@_first_marks_tl
+\tl_new:N \l_@@_last_marks_tl
+%    \end{macrocode}
+%  \end{macro}
+%
+%
+%  \begin{macro}{mc@reinsert@marks}
+%    So reinserting it just means executing the token lists (with some
+%    surrounding debugging statements).
+%    \begin{macrocode}
+\def\mc@reinsert@marks{
+  \@@_debug_marks:n
+     { \typeout{Marks:~ --~ reinsert~ marks~ (multicol)} }
+     \l_@@_first_marks_tl \l_@@_last_marks_tl
+  \@@_debug_marks:n
+     { \typeout{Marks:~ --~ finished~ reinserting~ marks~ (multicol)} }
+}
+%    \end{macrocode}
+%  \end{macro}
+%
+%
+% \subsubsection{Interfaces used in the \mc{} code and its output routines}
+%
+%
+%  \begin{macro}{\mc@prepare@mark@regions}
+%    When a \mc{} environment starts we need to clear the column
+%    region as it it may contain quite old data. Otherwise, that data
+%    would be used to generate the top marks and that would be
+%    obviously wrong for the first column.
+%    \begin{macrocode}
+\def \mc@prepare@mark@regions {
+%    \end{macrocode}
+%    However, before we do this we need to save away the current
+%    column data in case this is a boxed multicol, because when that
+%    finishes we need to restore the previous state --- this is not
+%    necessary for a normal \mc{} environment, because there the
+%    \texttt{column} region is overwritten in the standard output
+%    routine by copying the \texttt{page} region.
+%    \begin{macrocode}
+  \if@boxedmulticols
+    \__mark_copy_structure:nn{saved-column}{column}
+  \fi
```

Comment:
Why not \cs_new:Npn (or `protected`), and in `expl3` I'd expect to see `\legacy_if:nTF` here.

Reply:
well ..., partly because that is direct code from the other part of multicol, eg \mc@prepare@mark@regions is not a expl3 name either. Anyway, changed :-)